### ISSUE TITLE
Figure out how to handle secrets

### ISSUE THREAD

**Issue by bgrant0607** [kptdev/kpt#3125]
Forking the discussion from https://github.com/GoogleContainerTools/kpt/issues/2433#issuecomment-931896466

For usage in applications, there's https://secrets-store-csi-driver.sigs.k8s.io/. 

For cases where we do need to copy into Kubernetes Secrets, I'd recommend using a similar approach as GitOps, which is copying from the storage system, the secret manager in this case rather than git. In the GitOps layer that could be a special kind of sync request. In the kpt CLI we'll need to figure out how we want to stage secrets before the apply, and ensure we don't leak them.

There's also the issue of generation, for cases where users do want to pre-generate secrets and push them along with their other configuration.

**Comment by bgrant0607** [comment]
Yes, definitely as much as is practical we should recommend automatically rotated secrets: service accounts, workload identity, managed certs, and so on.

**Comment by bgrant0607** [comment]
https://dnastacio.medium.com/why-you-should-avoid-sealed-secrets-in-your-gitops-deployment-e50131d360dd
mentions https://external-secrets.io/

**Comment by yuwenma** [comment]
An example discovered in [Ghost Application package](https://github.com/GoogleContainerTools/kpt/pull/3381/commits/136f777959d49b2947f34232f2636e8127de75db):
 
Two secrets are used. One for ghost-app and the other is for MariaDB. 

The secretes are referred by ghost-app `Deployment` and mariadb `StatefulSet` on `spec.template.spec.containers[].env[].valueFrom.secretKeyRef` fields. 

The secretes are also injected in these two resources' `spec.template.metadata.annotations[].checksum/secrets`. checksum/secrets is a helm feature for [auto roll deployment](https://helm.sh/docs/howto/charts_tips_and_tricks/#automatically-roll-deployments). This ghost app is originally converted from a helm charts `bitnami/ghost`

**Comment by bgrant0607** [comment]
For completeness, kustomize supports secret generation plugins.
https://github.com/kubernetes-sigs/kustomize/blob/master/examples/secretGeneratorPlugin.md

Example plugin:
https://github.com/viaduct-ai/kustomize-sops

**Comment by bgrant0607** [comment]
External secrets was inducted into the CNCF sandbox:
https://lists.cncf.io/g/cncf-toc/message/7273

**Comment by bgrant0607** [comment]
We probably want a mechanism to help users not commit Secrets to git.
